torch/nativert/graph/GraphSignature.cpp

```diff
@@ -0,0 +1,505 @@
+    case torch::_export::Argument::Tag::AS_OPTIONAL_TENSORS:
+    case torch::_export::Argument::Tag::AS_SYM_BOOL:
+    case torch::_export::Argument::Tag::AS_SYM_BOOLS:
+    case torch::_export::Argument::Tag::AS_SYM_INT:
+    case torch::_export::Argument::Tag::AS_SYM_INTS:
+    case torch::_export::Argument::Tag::AS_SYM_FLOAT:
+    case torch::_export::Argument::Tag::AS_SYM_FLOATS:
+    case torch::_export::Argument::Tag::AS_CUSTOM_OBJ:
+      return true;
+    default:
+      return false;
+  }
+}
+
+std::pair<std::string, std::string> getSpecDetails(
+    const torch::_export::InputSpec& inputSpec) {
+  // Retrieve the argument name and spec tag name
+  std::string argName;
+  std::string tagName;
+  switch (inputSpec.tag()) {
+    case torch::_export::InputSpec::Tag::PARAMETER:
+      argName = inputSpec.get_parameter().get_arg().get_name();
+      tagName = "PARAMETER";
+      break;
+    case torch::_export::InputSpec::Tag::BUFFER:
+      argName = inputSpec.get_buffer().get_arg().get_name();
+      tagName = "BUFFER";
+      break;
+    case torch::_export::InputSpec::Tag::TENSOR_CONSTANT:
+      argName = inputSpec.get_tensor_constant().get_arg().get_name();
+      tagName = "TENSOR_CONSTANT";
+      break;
+    case torch::_export::InputSpec::Tag::CUSTOM_OBJ:
+      tagName = "CUSTOM_OBJ";
+      argName = inputSpec.get_custom_obj().get_arg().get_name();
+      break;
+    case torch::_export::InputSpec::Tag::USER_INPUT:
+      tagName = "USER_INPUT";
+      if (inputSpec.get_user_input().get_arg().tag() ==
+          torch::_export::Argument::Tag::AS_TENSOR) {
+        argName =
+            inputSpec.get_user_input().get_arg().get_as_tensor().get_name();
+      } else if (
+          inputSpec.get_user_input().get_arg().tag() ==
+          torch::_export::Argument::Tag::AS_CUSTOM_OBJ) {
+        argName =
+            inputSpec.get_user_input().get_arg().get_as_custom_obj().get_name();
+      } else {
+        TORCH_CHECK(false, "Unsupported USER_INPUT argument type.");
+      }
+      break;
+    case torch::_export::InputSpec::Tag::CONSTANT_INPUT:
+      argName = inputSpec.get_constant_input().get_name();
+      tagName = "CONSTANT_INPUT";
+      break;
+    case torch::_export::InputSpec::Tag::TOKEN:
+      TORCH_CHECK(false, "Token inputs not implemented yet.");
+    default:
+      TORCH_CHECK(false, "Unknown InputSpec tag encountered.");
+  }
+  return std::make_pair(argName, tagName);
+}
+
+void checkInputOrders(
+    const std::vector<torch::_export::InputSpec>& inputSpecs) {
+  // Map each tag to its index in the expected order
+  std::unordered_map<torch::_export::InputSpec::Tag, size_t> tagOrderMap = {
+      {torch::_export::InputSpec::Tag::TOKEN, 0},
+      {torch::_export::InputSpec::Tag::PARAMETER, 1},
+      {torch::_export::InputSpec::Tag::BUFFER, 2},
+      {torch::_export::InputSpec::Tag::TENSOR_CONSTANT, 3},
+      {torch::_export::InputSpec::Tag::CUSTOM_OBJ, 4}};
```

Comment:
1) creating a new hash table on each call to this function is inefficient
2) this is very small; linear search on a static constexpr array of std::pair would very likely be fine
3) you don't need to use a whole size_t for the value. uint32_t would be fine